opentracing-impl-java8/src/main/java/io/opentracing/propagation/TextFormatExtractorImpl.java

```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.opentracing.propagation;
+
+import java.util.Map;
+
+import io.opentracing.Tracer;
+
+final class TextFormatExtractorImpl implements TextFormatExtractor {
+
+    /** How the trace state is serialised and can be identified separately to baggage during deserialisation,
+     * for the tracing implementation being used, is contextual to the extraction process.
+     */
+    interface TraceState {
+        /** Returns true if the key+value belongs in traceState, otherwise it is presumed to be baggage. */
+        boolean isTraceState(String key, Object value);
+    }
+
+    private final TraceState traceState;
+
+    TextFormatExtractorImpl(TraceState traceState) {
+        this.traceState = traceState;
+    }
+
+    @Override
+    public Tracer.SpanBuilder join(Map carrier) {
+        AbstractTracer.SpanBuilderImpl builder = new AbstractTracer.SpanBuilderImpl();
+        for (Map.Entry<String, String> entry : ((Map<String, String>) carrier).entrySet()) {
+            if (traceState.isTraceState(entry.getKey(), entry.getValue())) {
+                builder.traceState.put(entry.getKey(), entry.getValue());
+            } else {
```

Comment:
it needs to do a similar check for `isBaggage`, because the carrier is allowed to contain more headers than what the sending tracer injected (essentially Map is an abstraction of all HTTP headers from the request, not just tracing related)